### PR TITLE
Improve types for `test.each`, `describe.each`

### PR DESCRIPTION
**Pull request by colinhacks**
Per https://discord.com/channels/876711213126520882/1153238257853800468/1153238257853800468

### CODE REVIEW

**Review by paperclover — APPROVED** [review]

File: packages/bun-types/bun-test.d.ts [hunk 1]
```diff
@@ -174,18 +174,18 @@ declare module "bun:test" {
...
+    each<T extends Readonly<Readonly<[any, ...any[]]>[]>>(
+      table: T,
     ): (
       label: string,
-      fn: (...args: T) => void | Promise<unknown>,
+      fn: (...args: Readonly<T>[number]) => void | Promise<unknown>,
       options?: number | TestOptions,
     ) => void;
-    each<T>(
+    each<T extends Array<any>>(
```
Comment (mProjectsCode) [inline packages/bun-types/bun-test.d.ts:184]: This breaks support for `each` with an array of objects. Now only an array of arrays is supported.

File: packages/bun-types/bun-test.d.ts [hunk 2]
```diff
@@ -419,18 +419,18 @@ declare module "bun:test" {
...
       table: ReadonlyArray<T>,
     ): (
       label: string,
-      fn: (arg: T, done: (err?: unknown) => void) => void | Promise<unknown>,
```
Comment (mProjectsCode) [inline packages/bun-types/bun-test.d.ts:433]: What happened to the `done` callback?

File: packages/bun-types/bun-test.d.ts [hunk 3]
```diff
@@ -419,18 +419,18 @@ declare module "bun:test" {
...
+    each<T extends Readonly<Readonly<[any, ...any[]]>[]>>(
+      table: T,
     ): (
       label: string,
-      fn: (...args: T) => void | Promise<unknown>,
+      fn: (...args: Readonly<T>[number]) => void | Promise<unknown>,
       options?: number | TestOptions,
     ) => void;
-    each<T>(
+    each<T extends Array<any>>(
```
Comment (mProjectsCode) [inline packages/bun-types/bun-test.d.ts:429]: Same thing as above

**Review by mProjectsCode — COMMENTED** [review]
This is a good change, but it breaks support for `each` with an array of objects and the `done` callback was lost along the way.

**Comment by mProjectsCode** [comment]
@colinhacks @paperdave I would suggest taking another look at this.

File: packages/bun-types/bun-test.d.ts (hunk 2 above)
Comment (colinhacks) [reply packages/bun-types/bun-test.d.ts:433]: The previous typings that included `done` were just completely wrong. There was no guarantee that `done` would be the second argument to this. It's basically impossible to include `done` in the types in the general case here, if we want any sort of inference on the argument to `.each()`.

**Pull request merged**